fishtank/src/backend/docker.ts

```diff
@@ -0,0 +1,88 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+import { ChildProcess, execFile, ExecFileOptions } from 'child_process'
+import { promisify } from 'util'
+
+const DEFAULT_COMMAND = 'docker'
+
+type ExecFilePromiseReturn = { child: ChildProcess; stdout: string; stderr: string }
+
+type ExecFilePromiseError = {
+  cmd: string
+  code: number | string
+  stdout: string
+  stderr: string
+}
+
+const execChildProcess = promisify<
+  string,
+  readonly string[],
+  ExecFileOptions,
+  ExecFilePromiseReturn
+>(execFile)
+
+/* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access */
+const isExecFilePromiseError = (e: any): e is ExecFilePromiseError => {
+  return (
+    !!e &&
+    typeof e === 'object' &&
+    typeof e['cmd'] === 'string' &&
+    (typeof e['code'] === 'number' || typeof e['code'] === 'string') &&
+    typeof e['stdout'] === 'string' &&
+    typeof e['stderr'] === 'string'
+  )
+}
+
+export class DockerError extends Error {
+  constructor(
+    public readonly cmd: string,
+    public readonly exitCode: number | string,
+    public stdout: string,
+    public stderr: string,
+  ) {
+    super(`Command '${cmd}' exited with status ${exitCode}`)
+  }
+}
+
+export class Docker {
+  private readonly command: string
+
+  constructor(options?: { command?: string }) {
+    this.command = options?.command ?? DEFAULT_COMMAND
+  }
+
+  private async cmd(
+    args: readonly string[],
+    options: ExecFileOptions,
+  ): Promise<ExecFilePromiseReturn> {
+    try {
+      return await execChildProcess(this.command, args, options)
```

Comment:
How do we know if the node is started successfully? If there is any error message during the starting the node, eg `Error while fetching verified assets` in `ironfish start`, how do you plan to identify it and handle it?

Reply:
It's the job of the testers to ensure that the spawned nodes reach an intended state. The simulator does not have enough context to know what state is expected. Some test scenarios may expect all nodes to be inter-connected and synced, some may expect a few nodes to be out-of-sync, some may expect a few nodes to fail to start completely.

Once we start writing test scenarios, we may add some helper methods that we see fit to aid with the most common expectations, but at its core, the simulator cannot know what is considered a "success" as it needs to handle both happy and unhappy scenarios.

Reply:
Okay it makes sense the docker does not need to know the node. How will testers know if the node is not started successfully or identify the state of the node? The tester will interact with Docker class right? eg. tail the logs of nodes
Or do you have other plans for tester to interact with node?

Reply:
I think the `status` API is what people will rely on the most. Example: if a tester considers a "successful start" the fact that the node is synced, they can query the `status` of the node through the RPC interface and see if the node is actually synced.

Reply:
So in short the RPC should be the source of information wrt the status of the node, more than the logs

Reply:
What you described makes sense for the happy case. Yeah, if you see status with syncing in progress, it mean node started successfully. But what about error cases and edge case? eg. when status shows blockchain is not sync, how would user know the error message to debug? ( This is more common for CLI use case, I think it might be applicable if testings are failing and people debug it)

Reply:
If a human wants to debug an unexpected error, they can use `docker attach` to see the logs, or use the Iron Fish command line client to retrieve the information they need to debug.

If an automated test cases expects a node to fail, and wants to check whether it's correctly failing, then again RPC is the answer. If, for example, you want to have a node to fail syncing, you can use retrieve the status of the node through the RPC and verify that the node is out-of-sync.

Reply:
> If a human wants to debug an unexpected error, they can use docker attach to see the logs, or use the Iron Fish 
 command line client to retrieve the information they need to debug.

I don't quite agree with your workaround for this problem. First of all, the whole thing in the simulator can be done via docker cmd and ironfish cmd. The purpose of simulator is to make the process easier and efficient for developers. 
Secondly, the ironfish cli `logs` wont show any previous logs message. You can not soley rely on that to debug.

I think we should provide the ability to show logs from ironfish cmd in simulator CLI, have a function like `runAttached` or `attach` to see the logs. But this problem def does not have be resolved in this PR. But we want to have the discussion so we are on the same page.

Reply:
> If an automated test cases expects a node to fail, and wants to check whether it's correctly failing, then again RPC is the answer. If, for example, you want to have a node to fail syncing, you can use retrieve the status of the node through the RPC and verify that the node is out-of-sync.
This is not the use case I am thinking. Say if an automated test case is expected to pass, but when the test case runs with the latest IronFish version, it somehow fails. You want to debug what cause the failure. Getting the status will help indirectly in some case, but it will not reveal the failure message. For the case when a node is hanging, it does not give you the info to identify the root cause. 
As for the `logs` cmd, it does not show previous messages. So when you started a node and its hanging, running `logs` wont give you anything because hanging node does not produce any new logs.

Reply:
That can be a good idea. How about you write up a proposal and then we review it once it's ready?

Reply:
If you agree,  this requirement can be combined with the miner logs.  The project requirement need to update to include these and cover it in the spec. Let's first agree upon this and then create the tasks and then decide who will be working on it.